h5pt.go

```diff
@@ -12,237 +12,274 @@ package hdf5
+		enc.buf = append(enc.buf, raw[:]...)
 
 	case reflect.Float64:
-		ad.numRecords = 1
-		val := C.double(rv.Float())
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 8
+		enc.numRecords = 1
+		var raw [8]byte
+		nativeEndian.PutUint64(raw[:], math.Float64bits(rv.Float()))
+		enc.buf = append(enc.buf, raw[:]...)
+		enc.offset += 8
 
 	case reflect.Bool:
-		ad.numRecords = 1
-		val := C.uchar(0)
+		enc.numRecords = 1
+		val := byte(0)
 		if rv.Bool() {
 			val = 1
 		}
-		dataPtr = unsafe.Pointer(&val)
-		dataSize = 1
+		enc.buf = append(enc.buf, val)
+		enc.offset++
 
 	default:
 		return fmt.Errorf("hdf5: PT Append does not support datatype (%s).", rt.Kind())
 	}
 
-	ad.memSize = dataSize + ad.offset
-	ad.ptr = C.realloc(ad.ptr, ad.memSize)
-	C.memcpy(unsafe.Pointer(uintptr(ad.ptr)+uintptr(ad.offset)), dataPtr, dataSize)
-	ad.offset += dataSize
-
 	return nil
 }
 
+// Table is an hdf5 packet-table.
+type Table struct {
+	Identifier
+}
+
+func newPacketTable(id C.hid_t) *Table {
+	t := &Table{Identifier{id}}
+	runtime.SetFinalizer(t, (*Table).finalizer)
+	return t
+}
+
+func (t *Table) finalizer() {
+	if err := t.Close(); err != nil {
+		panic(fmt.Errorf("error closing packet table: %s", err))
+	}
+}
+
+// Close closes an open packet table.
+func (t *Table) Close() error {
+	if t.id == 0 {
+		return nil
+	}
+	err := h5err(C.H5PTclose(t.id))
+	t.id = 0
+	return err
+}
+
+// IsValid returns whether or not an indentifier points to a packet table.
+func (t *Table) IsValid() bool {
+	return C.H5PTis_valid(t.id) >= 0
+}
+
+func (t *Table) Id() int {
+	return int(t.id)
+}
+
+// ReadPackets reads a number of packets from a packet table.
+func (t *Table) ReadPackets(start, nrecords int, data interface{}) error {
+	c_start := C.hsize_t(start)
+	c_nrecords := C.size_t(nrecords)
+	rv := reflect.Indirect(reflect.ValueOf(data))
+	rt := rv.Type()
+	c_data := unsafe.Pointer(nil)
+	switch rt.Kind() {
+	case reflect.Array:
+		if rv.Len() < nrecords {
+			panic(fmt.Errorf("not enough capacity in array (cap=%d)", rv.Len()))
+		}
+		c_data = unsafe.Pointer(rv.Index(0).UnsafeAddr())
+
+	case reflect.Slice:
+		if rv.Len() < nrecords {
+			panic(fmt.Errorf("not enough capacity in slice (cap=%d)", rv.Len()))
+		}
+		slice := (*reflect.SliceHeader)(unsafe.Pointer(rv.UnsafeAddr()))
+		c_data = unsafe.Pointer(slice.Data)
+
+	default:
+		panic(fmt.Errorf("unhandled kind (%s), need slice or array", rt.Kind()))
+	}
+	err := C.H5PTread_packets(t.id, c_start, c_nrecords, c_data)
+	return h5err(err)
+}
+
 // Append appends packets to the end of a packet table.
 //
 // Struct values must only have exported fields, otherwise Append will panic.
 func (t *Table) Append(data interface{}) error {
-	var ad appendData
-	defer ad.free()
+	enc := &encoder{}
 
-	if err := extractValues(&ad, data); err != nil {
+	if err := enc.encode(data); err != nil {
 		return err
 	}
 
-	return h5err(C.H5PTappend(t.id, ad.numRecords, ad.ptr))
+	cData := unsafe.Pointer(&enc.buf[0])
```

Comment:
we should probably check for the length of that buffer before doing this, at least to send the user a legible or more interesting error, in case that buffer is `nil` or empty.